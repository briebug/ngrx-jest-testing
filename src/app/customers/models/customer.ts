export interface Customer {
  id: number;
  name: string;
}
