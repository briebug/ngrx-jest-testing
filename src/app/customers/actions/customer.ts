import { Action } from '@ngrx/store';
import { Customer } from '../models/customer';

export enum CustomerActionTypes {
  Load = '[Customer] Load',
  LoadSuccess = '[Customer] Load Success',
  LoadFail = '[Customer] Load Fail',
  Select = '[Customer] Select'
}

export class Load implements Action {
  readonly type = CustomerActionTypes.Load;
}

export class LoadSuccess implements Action {
  readonly type = CustomerActionTypes.LoadSuccess;

  constructor(public payload: Customer[]) {}
}

export class LoadFail implements Action {
  readonly type = CustomerActionTypes.LoadFail;

  constructor(public error: any) {}
}

export class Select implements Action {
  readonly type = CustomerActionTypes.Select;

  constructor(public payload: Customer) {}
}

export type CustomerActions = Load | LoadSuccess | LoadFail | Select;
