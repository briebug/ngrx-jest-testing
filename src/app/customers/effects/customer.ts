import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import {
  CustomerActionTypes,
  Load,
  LoadSuccess,
  LoadFail
} from '../actions/customer';

import { Customer } from '../models/customer';
import { map, switchMap, catchError } from 'rxjs/operators';
import { CustomersService } from '../services/customers';

@Injectable()
export class CustomerEffects {
  @Effect()
  load$: Observable<Action> = this.actions$.pipe(
    ofType<Load>(CustomerActionTypes.Load),
    switchMap(query => {
      return this.customersService
        .load()
        .pipe(
          map((customers: Customer[]) => new LoadSuccess(customers)),
          catchError(err => of(new LoadFail(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private customersService: CustomersService
  ) {}
}
