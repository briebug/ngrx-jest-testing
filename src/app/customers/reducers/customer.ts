import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { Customer } from '../models/customer';
import { CustomerActions, CustomerActionTypes } from '../actions/customer';

export interface State extends EntityState<Customer> {
  selectedCustomerId: number | null;
  loading: boolean;
  error: any;
}

export const adapter: EntityAdapter<Customer> = createEntityAdapter<Customer>();

export const initialState: State = adapter.getInitialState({
  selectedCustomerId: null,
  loading: false,
  error: null
});

export function reducer(state = initialState, action: CustomerActions): State {
  switch (action.type) {
    case CustomerActionTypes.Load:
      return { ...state, loading: true, selectedCustomerId: null };

    case CustomerActionTypes.LoadSuccess: {
      return { ...adapter.addMany(action.payload, state) };
    }

    case CustomerActionTypes.LoadFail: {
      return { ...state, error: { message: 'Error loading customers' } };
    }

    case CustomerActionTypes.Select: {
      return {
        ...state,
        selectedCustomerId: action.payload.id
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedCustomerId;
export const getLoading = (state: State) => state.loading;
export const getError = (state: State) => state.error;
