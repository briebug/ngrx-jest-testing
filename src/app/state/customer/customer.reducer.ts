import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { Customer } from './customer.model';
import { CustomerActions, CustomerActionTypes } from './customer.actions';

export interface State extends EntityState<Customer> {
  selectedCustomerId: number;
  loading: boolean;
  error: string;
}

export const adapter: EntityAdapter<Customer> = createEntityAdapter<Customer>();

export const initialState: State = adapter.getInitialState({
  selectedCustomerId: null,
  loading: false,
  error: null
});

export function reducer(state = initialState, action: CustomerActions): State {
  switch (action.type) {
    case CustomerActionTypes.Load:
      return { ...state, loading: true, selectedCustomerId: null };

    case CustomerActionTypes.LoadSuccess: {
      return { ...adapter.addMany(action.payload, state) };
    }

    case CustomerActionTypes.LoadFail: {
      console.log(action.error);
      return { ...state, error: 'Error loading customers' };
    }

    case CustomerActionTypes.Select: {
      return {
        ...state,
        selectedCustomerId: action.payload.id
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedCustomerId;
export const getLoading = (state: State) => state.loading;
export const getError = (state: State) => state.error;
