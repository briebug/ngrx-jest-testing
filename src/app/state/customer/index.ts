import { createSelector, createFeatureSelector } from '@ngrx/store';

import * as fromCustomers from './customer.reducer';
import { State as CustomersState } from './customer.reducer';

export const reducers = {
  customers: fromCustomers.reducer
};


export const getCustomersState = createFeatureSelector<CustomersState>(
  'customer'
);

export const {
  selectAll: getAllCustomers,
  selectEntities: getCustomerEntities,
  selectIds: getCustomerIds,
  selectTotal: getTotalCustomers
} = fromCustomers.adapter.getSelectors(getCustomersState);

export const getSelectedCustomerId = createSelector(
  getCustomersState,
  fromCustomers.getSelectedId
);

export const getSelectedCustomer = createSelector(
  getCustomerEntities,
  fromCustomers.getSelectedId,
  (entities, id) => entities[id]
);

export const getLoading = createSelector(
  getCustomersState,
  fromCustomers.getLoading
);

export const getError = createSelector(
  getCustomersState,
  fromCustomers.getError
);
