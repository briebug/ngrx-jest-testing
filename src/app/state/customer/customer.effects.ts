import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';
import { map, switchMap, catchError } from 'rxjs/operators';

import {
  CustomerActionTypes,
  Load,
  LoadSuccess,
  LoadFail
} from './customer.actions';

import { Customer } from './customer.model';
import { CustomersService } from '@core/services/customer.service';

@Injectable()
export class CustomerEffects {
  @Effect()
  load$: Observable<Action> = this.actions$.pipe(
    ofType<Load>(CustomerActionTypes.Load),
    switchMap(query => {
      return this.customersService
        .load()
        .pipe(
          map((customers: Customer[]) => new LoadSuccess(customers)),
          catchError(err => of(new LoadFail(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private customersService: CustomersService
  ) {}
}
