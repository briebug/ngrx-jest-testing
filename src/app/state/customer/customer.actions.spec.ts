import * as CustomerActions from './customer.actions';
import { Customer } from './customer.model';

describe('Customer Actions', () => {
  const customer = <Customer>{ id: 1, name: 'Test User' };

  describe('Load Customers', () => {
    test('should create Customer Load Action', () => {
      const action = new CustomerActions.Load();

      expect(action.type).toMatchSnapshot();
    });
  });

  describe('Load Customers Success', () => {
    test('should create Customer Load Success Action', () => {
      const payload = <Customer[]>[customer],
        action = new CustomerActions.LoadSuccess(payload);

      expect(action.type).toMatchSnapshot();
      expect(action.payload).toMatchSnapshot();
    });
  });

  describe('Load Customers Success', () => {
    test('should create Customer Load Fail Action', () => {
      const error = { status: 1, error: 'Error loading customer' },
        action = new CustomerActions.LoadFail(error);

      expect(action.type).toMatchSnapshot();
      expect(action.error).toMatchSnapshot();
    });
  });

  describe('Customer Select', () => {
    test('should set payload to customer', () => {
      const action = new CustomerActions.Select(customer);

      expect(action.type).toMatchSnapshot();
      expect(action.payload).toMatchSnapshot();
    });

    test('should set customer to payload in snapshot', () => {
      const action = new CustomerActions.Select(customer);

      expect(action.payload).toMatchSnapshot();
    });
  });
});
