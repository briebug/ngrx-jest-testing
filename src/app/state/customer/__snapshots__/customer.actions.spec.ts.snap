// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer Actions Customer Select should set customer to payload in snapshot 1`] = `
Object {
  "id": 1,
  "name": "Test User",
}
`;

exports[`Customer Actions Customer Select should set payload to customer 1`] = `"[Customer] Select"`;

exports[`Customer Actions Customer Select should set payload to customer 2`] = `
Object {
  "id": 1,
  "name": "Test User",
}
`;

exports[`Customer Actions Load Customers Success should create Customer Load Fail Action 1`] = `"[Customer] Load Fail"`;

exports[`Customer Actions Load Customers Success should create Customer Load Fail Action 2`] = `
Object {
  "error": "Error loading customer",
  "status": 1,
}
`;

exports[`Customer Actions Load Customers Success should create Customer Load Success Action 1`] = `"[Customer] Load Success"`;

exports[`Customer Actions Load Customers Success should create Customer Load Success Action 2`] = `
Array [
  Object {
    "id": 1,
    "name": "Test User",
  },
]
`;

exports[`Customer Actions Load Customers should create Customer Load Action 1`] = `"[Customer] Load"`;
